File: src/players.py
import json
import re
import time
from typing import List, Dict, Any
import os.path

from api import call_api

DIR = os.path.dirname(__file__)

PLAYERS_CACHE_FILE = os.path.join(DIR, "..", "cache", "pcache.json")
PLAYERS_FILE = os.path.join(DIR, "frontend", "data", "players.json")


def get_players_data(players: List[str]) -> Dict[str, Dict]:
    for player in players:
        with open(PLAYERS_CACHE_FILE, 'r') as f:
            output = json.load(f)
        if player not in output:
            wiki_text = call_api({
                "action": "parse",
                "prop": "wikitext",
                "page": player,
                "format": "json",
                "section": 0,
            })
            print("called api for", player)
            output[player] = wiki_text['parse']
            with open(PLAYERS_CACHE_FILE, 'w') as f:
                json.dump(output, f, indent=2)
            time.sleep(30)
        else:
            print("didn't need api for", player)

    with open(PLAYERS_CACHE_FILE) as f:
        output = json.load(f)
    return output


DATE_RE = re.compile(r"[\w?]{4}-[\w?]{2}-[\w?]{2}")


def process_players_data(output: Dict[str, Dict]):
    """
    Go through the wikitext output and process it into a clean JSON format
    """
    processed = []
    for player in output:
        lines: List[str] = output[player]["wikitext"]["*"].split('\n')

        player = {
            "events": []
        }
        in_infobox = False
        in_history = False
        for line in lines:
            if line.startswith('{{Infobox'):
                in_infobox = True
            elif in_infobox:
                if line.startswith('|id='):
                    player["name"] = line.replace('|id=', '')

            if not in_history and line.startswith('|history'):
                in_history = True
            elif in_history:
                if not line.startswith('{{TH'):
                    break

                parts = line.replace('{{', '').replace('}}', '').split('|')
                dates = parts[1].split(' ')
                event = {
                    "start": dates[0],
                    "team": parts[2],
                }
                if re.match(DATE_RE, dates[-1]):
                    event["end"] = dates[-1]
                player["events"].append(event)

        processed.append(player)

    with open(PLAYERS_FILE, 'w') as f:
        json.dump(processed, f, indent=2)

    return processed

File: src/data/sample/players.ts
import { Player } from "../../types";

export const PLAYERS = [
  { memberships: [{ join: "2021-01-01", team: "Team 1" }], name: "Player 1" },
  { memberships: [{ join: "2021-01-01", team: "Team 1" }], name: "Player 2" },
  {
    memberships: [
      { join: "2021-01-01", leave: "2021-02-15", team: "Team 1" },
      { join: "2021-02-16", team: "Team 2" },
    ],
    name: "Player 3",
  },
  {
    memberships: [
      { join: "2021-01-01", leave: "2021-02-17", team: "Team 2" },
      { join: "2021-02-20", team: "Team 1" },
    ],
    name: "Player 4",
  },
  { memberships: [{ join: "2021-01-05", team: "Team 2" }], name: "Player 5" },
  { memberships: [{ join: "2021-01-06", team: "Team 2" }], name: "Player 6" },
] as Player[];
